### PR TITLE
Added queuing api

### PR DESCRIPTION
**Pull request by sliverc**
Added queue api and adjusted critical api calls to be queued 

Task on queue have there own output. This is kept to a minimum at this point and might be extended on a need basis.

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -31,29 +35,66 @@ const (
...
+func pushToQueue(name string, proc func(out *task.Output) error) task.Task {
+
+	return context.Queue().Push(name, func(out *task.Output) error {
+		err := acquireDatabaseConnection()
+
+		if err != nil {
+			return err
+		}
+
+		defer releaseDatabaseConnection()
```
Comment (winged) [inline api/api.go:93]: This should be moved to before the if block in my understanding

File: task/queue.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+// GetTasks gets complete list of tasks
+func (q *Queue) GetTasks() []Task {
+	var tasks []Task
+	q.Lock()
+	for _, task := range q.tasks {
+		tasks = append(tasks, *task)
+	}
+
+	q.Unlock()
```
Comment (winged) [inline task/queue.go:73]: Shouldn't this be called via defer?

File: task/queue.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+func (q *Queue) GetTaskByID(ID int) (Task, error) {
+	q.Lock()
+	tasks := q.tasks
+	q.Unlock()
+
+	for _, task := range tasks {
+		if task.ID == ID {
+			return *task, nil
+		}
+	}
```
Comment (winged) [inline task/queue.go:87]: I assume linear lookup is fine here as we're not expecting too many tasks... However maybe it'd be a good idea still to store them in a by-id hashmap/similar?

**Review by winged — CHANGES_REQUESTED** [review]

File: task/queue.go (hunk 2 above)
Comment (sliverc) [reply task/queue.go:73]: defer would work but it this block no errors can happen and therefore unlock will be happen earlier this way

File: task/queue.go (hunk 3 above)
Comment (sliverc) [reply task/queue.go:87]: There are lots of other code blocks in aptly where a for loop happens the same way (e.g. looping over snapshots etc.).

In this case I think it shouldn't be a problem and would leave it for now. If there performance issues all those loops would need to be addressed.

File: api/api.go (hunk 1 above)
Comment (sliverc) [reply api/api.go:93]: if acquiring of database connection is not successful, releasing such connection is not necessary. Therefore this defer is after the first error block.

**Pull request merged**